tests/std/tests/P0912R5_coroutine/test.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <version> // TRANSITION, P0912R5 Library Support For Coroutines
+#if defined(__cpp_lib_coroutine) && __cpp_lib_coroutine >= 201902L // TRANSITION, P0912R5 Library Support For Coroutines
+
+#include <assert.h>
+#include <coroutine>
+#include <exception>
+using namespace std;
+
+int g_tasks_destroyed{0};
+
+struct Task {
+    struct Promise {
+        int result{-1000};
+        coroutine_handle<> previous;
+
+        Task get_return_object() {
+            return {*this};
+        }
+
+        suspend_always initial_suspend() {
+            return {};
+        }
+
+        auto final_suspend() noexcept {
+            struct Awaiter {
+                bool await_ready() noexcept {
+                    return false;
+                }
+
+                void await_resume() noexcept {}
+
+                coroutine_handle<> await_suspend(coroutine_handle<Promise> h) {
+                    if (auto& pre = h.promise().previous; pre) {
+                        return pre; // resume awaiting coroutine
+                    }
+
+                    // If there is no previous coroutine to resume, we've reached the outermost coroutine.
+                    // Return a noop coroutine to allow control to return back to the caller.
+                    return noop_coroutine();
+                }
+            };
+
+            return Awaiter{};
+        }
+
+        void return_value(const int v) {
+            result = v;
+        }
+
+        void unhandled_exception() noexcept {
+            terminate();
+        }
+    };
+
+    using promise_type = Promise;
+
+    bool await_ready() const noexcept {
+        return false;
+    }
+
+    int await_resume() {
+        return coro.promise().result;
+    }
+
+    auto await_suspend(coroutine_handle<> enclosing) {
+        coro.promise().previous = enclosing;
+        return coro; // resume ourselves
+    }
+
+    Task(Task&& rhs) noexcept : coro(rhs.coro) {
+        rhs.coro = nullptr;
+    }
+
+    Task(Promise& p) : coro(coroutine_handle<Promise>::from_promise(p)) {}

```

Comment:
The construction appeared to be implicit, but I don't understand coroutines well enough to definitively answer this (which is why I refrained from altering this part of the code). @joemmett?

Reply:
The only construction is the implicit one on line 20, which could be changed to agree. I'm being hyper-paranoid since this implicit constructor takes ownership, but to be fair there aren't so many `Promise&`s flying around in this test that someone is likely to accidentally create a double-deletion scenario.

Reply:
Somehow `triangular_number` is returning a `Task` by saying `co_return INT;` - I don't know what machinery is making that possible.